Add reference for @preconcurrency (#307)
apple/swift-book: TSPL.docc/ReferenceManual/Attributes.md

  because of the larger symbol table slowing dyld down.
-->

### preconcurrency

Apply this attribute to a declaration,
to suppress strict concurrency checking.

On an import declaration,
this attribute reduces the strictness of concurrency checking
for code that uses types from the imported module.
Specifically,
types from the imported module

> that aren't explicitly marked as either sendable or nonsendable

Reviewer: ```suggestion
that aren't explicitly marked as nonsendable
```

Sendable types can always be used in a context that requires a sendable type, so I think you can simplify the above a little.

can be used in a context that requires sendable types.

On other declarations,
this attribute reduces the strictness of concurrency checking
for code that uses the symbol being declared.

> When you use this symbol in a scope that has  minimal concurrency checking,

Reviewer 2: ```suggestion
When you use this symbol in a scope that has minimal concurrency checking,
```

Typo: double space.

Author: Thanks!  Fixed in b79a5992.

> When you use this symbol in a scope that has minimal concurrency checking,
> you can use members of this type
> that aren't explicitly marked as either sendable or nonsendable
> in a context that requires sendable types.

Reviewer: It's not really about members of this type. I think of it as not enforcing any of the concurrency-related constraints of the symbol (such as Sendable or a global actor) on uses of type. Let me take a stab at it.

```suggestion
When you use this symbol in a scope that has minimal concurrency checking,
concurrency-related constraints specified by that symbol (such as `Sendable` requirements or global actors) are not checked.
```

> in a context that requires sendable types.

Author: Tech reviewer: Please confirm.  I'm putting together facts from a few places in the SE proposal, and I'm not sure this is correct.  Elsewhere, it says that `@preconcurrency` on a declaration for a nominal type generates code that interoperates with Swift 5 binaries.

Reviewer 3: I don't think that's quite right. I think I would describe it more like this:

> On other declarations, `@preconcurrency` means that the declaration has been modified from its original version to add `Sendable` constraints or other concurrency-related attributes. Swift will generate code that is compatible with versions of the library that predated those additions, and it will not enforce the declaration's concurrency properties in scopes that have minimal concurrency checking.

You can apply this attribute
to the following kinds of declarations:

- Structures, classes, and actors
- Enumerations and enumeration cases
- Protocols
- Variables and constants
- Subscripts
- Initializers

> - Functions

Reviewer: Should imports be on this list, since they're mentioned above?

You can use this attribute as follows,
to aid in migrating code to strict concurrency checking:

1. Enable strict checking.
1. Annotate imports and declarations that need to be migrated
   with the `preconcurrency` attribute.
1. After migrating a declaration,

>    remove the `preconcurrency` attribute from it.

Reviewer: `@preconcurrency` on declarations has a different role than on imports. The flow you describe is the right one for imports, although I'd collapse the middle two steps:
1. Enable strict checking
2. Annotate imports with `@preconcurrency` for imports of libraries that haven't enabled strict checking
3. Once an imported library has enabled strict checking, remove `@preconcurrency` from the import. (The compiler will also produce a warning)

For a declaration, add `@preconcurrency` when you are adding concurrency-related constraints to the declaration but you still have clients that haven't migrated to Swift concurrency. Remove the `@preconcurrency` when all of your clients have migrated.
